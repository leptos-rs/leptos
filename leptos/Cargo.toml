[package]
name = "leptos"
version = { workspace = true }
edition = "2021"
authors = ["Greg Johnston"]
license = "MIT"
repository = "https://github.com/leptos-rs/leptos"
description = "Leptos is a full-stack, isomorphic Rust web framework leveraging fine-grained reactivity to build declarative user interfaces."
readme = "../README.md"

[dependencies]
any_spawner = { workspace = true, features = ["wasm-bindgen"] }
cfg-if = "1"
leptos_dom = { workspace = true }
leptos_macro = { workspace = true }
leptos_reactive = { workspace = true }
leptos_server = { workspace = true }
leptos_config = { workspace = true }
leptos-spin-macro = { git = "https://github.com/fermyon/leptos-spin", optional = true }
oco = { workspace = true }
paste = "1"
reactive_graph = { workspace = true, features = ["serde"] }
tachys = { workspace = true, features = ["oco", "reactive_graph"] }
tracing = "0.1"
typed-builder = "0.18"
typed-builder-macro = "0.18"
serde = { version = "1", optional = true }
serde_json = { version = "1", optional = true }
server_fn = { workspace = true, features = [
  "form-redirects",
  "browser",
  "url",
  "cbor",
] }
web-sys = { version = "0.3.63", features = [
  "ShadowRoot",
  "ShadowRootInit",
  "ShadowRootMode",
] }
wasm-bindgen = { version = "0.2" }

[features]
default = ["serde"]
csr = ["leptos_macro/csr", "leptos_reactive/csr", "leptos_server/csr"]
hydrate = [
  "leptos_macro/hydrate",
  "leptos_reactive/hydrate",
  "leptos_server/hydrate",
]
default-tls = ["leptos_server/default-tls", "server_fn/default-tls"]
rustls = ["leptos_server/rustls", "server_fn/rustls"]
ssr = [
  "leptos_macro/ssr",
  "leptos_reactive/ssr",
  "leptos_server/ssr",
  "server_fn/ssr",
]
nightly = [
  "leptos_dom/nightly",
  "leptos_macro/nightly",
  "leptos_reactive/nightly",
  "leptos_server/nightly",
  "tachys/nightly",
]
serde = ["leptos_reactive/serde"]
serde-lite = ["leptos_reactive/serde-lite"]
miniserde = ["leptos_reactive/miniserde"]
rkyv = ["leptos_reactive/rkyv", "server_fn/rkyv"]
tracing = [
  "reactive_graph/tracing",
] #, "leptos_macro/tracing", "leptos_dom/tracing"]
nonce = ["leptos_dom/nonce"]
spin = ["leptos_reactive/spin", "leptos-spin-macro"]
experimental-islands = [
  "leptos_dom/experimental-islands",
  "leptos_macro/experimental-islands",
  "leptos_reactive/experimental-islands",
  "dep:serde",
  "dep:serde_json",
]
trace-component-props = [
  "leptos_dom/trace-component-props",
  "leptos_macro/trace-component-props",
]
delegation = ["tachys/delegation"]

[package.metadata.cargo-all-features]
denylist = [
  "nightly",
  "tracing",
  "template_macro",
  "rustls",
  "default-tls",
  "wasm-bindgen",
  "rkyv",                  # was causing clippy issues on nightly
  "trace-component-props",
]
skip_feature_sets = [
  [
    "csr",
    "ssr",
  ],
  [
    "csr",
    "hydrate",
  ],
  [
    "ssr",
    "hydrate",
  ],
  [
    "serde",
    "serde-lite",
  ],
  [
    "serde-lite",
    "miniserde",
  ],
  [
    "serde",
    "miniserde",
  ],
  [
    "serde",
    "rkyv",
  ],
  [
    "miniserde",
    "rkyv",
  ],
  [
    "serde-lite",
    "rkyv",
  ],
  [
    "default-tls",
    "rustls",
  ],
]

[package.metadata.docs.rs]
rustdoc-args = ["--generate-link-to-definition"]
