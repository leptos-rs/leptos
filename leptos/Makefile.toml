extend = { path = "../cargo-make/main.toml" }

[tasks.check]
clear = true
dependencies = [
	"check-all",
	"check-wasm",
	"check-release",
	"check-wasm-release",
]

[tasks.check-wasm]
clear = true
dependencies = ["check-hydrate", "check-csr"]

[tasks.check-wasm-release]
clear = true
dependencies = ["check-hydrate-release", "check-csr-release"]

[tasks.check-hydrate]
command = "cargo"
args = [
	"check",
	"--no-default-features",
	"--features=hydrate",
	"--target=wasm32-unknown-unknown",
]

[tasks.check-csr]
command = "cargo"
args = [
	"check",
	"--no-default-features",
	"--features=csr",
	"--target=wasm32-unknown-unknown",
]

[tasks.check-hydrate-release]
command = "cargo"
args = [
	"check",
	"--release",
	"--no-default-features",
	"--features=hydrate",
	"--target=wasm32-unknown-unknown",
]

[tasks.check-csr-release]
command = "cargo"
args = [
	"check",
	"--release",
	"--no-default-features",
	"--features=csr",
	"--target=wasm32-unknown-unknown",
]

[tasks.check-release]
command = "cargo"
args = ["check", "--release"]
