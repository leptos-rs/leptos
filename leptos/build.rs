use rustc_version::{version_meta, Channel};

fn main() {
    let target = std::env::var("TARGET").unwrap_or_default();

    // Set cfg flags depending on release channel
    if matches!(version_meta().unwrap().channel, Channel::Nightly) {
        println!("cargo:rustc-cfg=rustc_nightly");
    }
    // Set cfg flag for getrandom wasm_js
    if target == "wasm32-unknown-unknown" {
        // Set a custom cfg flag for wasm builds
        println!("cargo:rustc-cfg=getrandom_backend=\"wasm_js\"");
    }
}
