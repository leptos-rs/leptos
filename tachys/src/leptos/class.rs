use crate::{html::class::IntoClass, renderer::DomRenderer};
use leptos_reactive::{create_render_effect, Effect};

impl<F, C, R> IntoClass<R> for F
where
    F: Fn() -> C + 'static,
    C: IntoClass<R> + 'static,
    C::State: 'static,
    R: DomRenderer,
    R::ClassList: 'static,
    R::Element: Clone + 'static,
{
    type State = Effect<C::State>;

    fn to_html(self, class: &mut String) {
        let value = self();
        value.to_html(class);
    }

    fn hydrate<const FROM_SERVER: bool>(self, el: &R::Element) -> Self::State {
        // TODO FROM_SERVER vs template
        let el = el.clone();
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.hydrate::<FROM_SERVER>(&el)
            }
        })
    }

    fn build(self, el: &R::Element) -> Self::State {
        let el = el.to_owned();
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.build(&el)
            }
        })
    }

    fn rebuild(self, state: &mut Self::State) {}
}

impl<F, R> IntoClass<R> for (&'static str, F)
where
    F: Fn() -> bool + 'static,
    R: DomRenderer,
    R::ClassList: 'static,
    R::Element: Clone,
{
    type State = Effect<bool>;

    fn to_html(self, class: &mut String) {
        let (name, f) = self;
        let include = f();
        if include {
            <&str as IntoClass<R>>::to_html(name, class);
        }
    }

    fn hydrate<const FROM_SERVER: bool>(self, el: &R::Element) -> Self::State {
        // TODO FROM_SERVER vs template
        let (name, f) = self;
        let class_list = R::class_list(el);
        create_render_effect(move |prev| {
            let include = f();
            if Some(include) != prev {
                if include {
                    R::add_class(&class_list, name);
                } else {
                    R::remove_class(&class_list, name);
                }
            }
            include
        })
    }

    fn build(self, el: &R::Element) -> Self::State {
        let (name, f) = self;
        let class_list = R::class_list(el);
        create_render_effect(move |prev| {
            let include = f();
            if Some(include) != prev {
                if include {
                    R::add_class(&class_list, name);
                } else {
                    R::remove_class(&class_list, name);
                }
            }
            include
        })
    }

    fn rebuild(self, state: &mut Self::State) {}
}
