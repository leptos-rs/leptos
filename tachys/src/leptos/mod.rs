use crate::{
    hydration::Cursor,
    renderer::Renderer,
    view::{
        Mountable, Position, PositionState, Render, RenderHtml, ToTemplate,
    },
};
use leptos_reactive::{create_render_effect, Effect, SignalDispose};

mod class;
mod style;

impl<F, V> ToTemplate for F
where
    F: Fn() -> V,
    V: ToTemplate,
{
    fn to_template(
        buf: &mut String,
        class: &mut String,
        style: &mut String,
        position: &mut Position,
    ) {
        // FIXME this seems wrong
        V::to_template(buf, class, style, position)
    }
}

impl<F, V, R> Render<R> for F
where
    F: Fn() -> V + 'static,
    V: Render<R>,
    V::State: 'static,
    R: Renderer,
{
    type State = Effect<V::State>;

    fn build(self) -> Self::State {
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.build()
            }
        })
    }

    #[track_caller]
    fn rebuild(self, state: &mut Self::State) {
        crate::log(&format!(
            "[REBUILDING EFFECT] Is this a mistake? {}",
            std::panic::Location::caller(),
        ));
        let old_effect = std::mem::replace(state, self.build());
        old_effect.dispose();
    }
}

impl<F, V, R> RenderHtml<R> for F
where
    F: Fn() -> V + 'static,
    V: RenderHtml<R>,
    V::State: 'static,
    R: Renderer + 'static,
    R::Node: Clone,
    R::Element: Clone,
{
    const MIN_LENGTH: usize = V::MIN_LENGTH;

    fn to_html_with_buf(self, buf: &mut String, position: &PositionState) {
        let value = self();
        value.to_html_with_buf(buf, position);
    }

    fn hydrate<const FROM_SERVER: bool>(
        self,
        cursor: &Cursor<R>,
        position: &PositionState,
    ) -> Self::State {
        let cursor = cursor.clone();
        let position = position.clone();
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.hydrate::<FROM_SERVER>(&cursor, &position)
            }
        })
    }
}

impl<M, R> Mountable<R> for Effect<M>
where
    M: Mountable<R> + 'static,
    R: Renderer,
{
    fn unmount(&mut self) {
        self.with_value_mut(|value| {
            if let Some(value) = value {
                value.unmount()
            }
        });
    }

    fn mount(
        &mut self,
        parent: &<R as Renderer>::Element,
        marker: Option<&<R as Renderer>::Node>,
    ) {
        self.with_value_mut(|value| {
            if let Some(state) = value {
                state.mount(parent, marker);
            }
        });
    }

    fn insert_before_this(
        &self,
        parent: &<R as Renderer>::Element,
        child: &mut dyn Mountable<R>,
    ) -> bool {
        self.with_value_mut(|value| {
            value
                .as_mut()
                .map(|value| value.insert_before_this(parent, child))
        })
        .flatten()
        .unwrap_or(false)
    }
}
/*
#[cfg(test)]
mod tests {
    use crate::{
        html::element::{button, main, HtmlElement},
        renderer::mock_dom::MockDom,
        view::Render,
    };
    use leptos_reactive::{create_runtime, RwSignal, SignalGet, SignalSet};

    #[test]
    fn create_dynamic_element() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> =
            button((), move || count.get().to_string());
        let el = app.build();
        assert_eq!(el.el.to_debug_html(), "<button>0</button>");
        rt.dispose();
    }

    #[test]
    fn update_dynamic_element() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> =
            button((), move || count.get().to_string());
        let el = app.build();
        assert_eq!(el.el.to_debug_html(), "<button>0</button>");
        count.set(1);
        assert_eq!(el.el.to_debug_html(), "<button>1</button>");
        rt.dispose();
    }

    #[test]
    fn update_dynamic_element_among_siblings() {
        let rt = create_runtime();
        let count = RwSignal::new(0);
        let app: HtmlElement<_, _, _, MockDom> = main(
            (),
            button(
                (),
                ("Hello, my ", move || count.get().to_string(), " friends."),
            ),
        );
        let el = app.build();
        assert_eq!(
            el.el.to_debug_html(),
            "<main><button>Hello, my 0 friends.</button></main>"
        );
        count.set(42);
        assert_eq!(
            el.el.to_debug_html(),
            "<main><button>Hello, my 42 friends.</button></main>"
        );
        rt.dispose();
    }
}
 */
