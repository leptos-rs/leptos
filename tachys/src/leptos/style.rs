use crate::{html::style::IntoStyle, renderer::DomRenderer};
use leptos_reactive::{create_render_effect, Effect};
use std::borrow::Cow;

impl<F, S, R> IntoStyle<R> for (&'static str, F)
where
    F: Fn() -> S + 'static,
    S: Into<Cow<'static, str>>,
    R: DomRenderer,
    R::CssStyleDeclaration: Clone + 'static,
{
    type State = Effect<(R::CssStyleDeclaration, Cow<'static, str>)>;

    fn to_html(self, style: &mut String) {
        let (name, f) = self;
        let value = f();
        style.push_str(name);
        style.push(':');
        style.push_str(&value.into());
        style.push(';');
    }

    fn hydrate<const FROM_SERVER: bool>(self, el: &R::Element) -> Self::State {
        let (name, f) = self;
        // TODO FROM_SERVER vs template
        let style = R::style(el);
        create_render_effect(move |prev| {
            let value = f().into();
            if let Some(mut state) = prev {
                let (style, prev): &mut (
                    R::CssStyleDeclaration,
                    Cow<'static, str>,
                ) = &mut state;
                if &value != prev {
                    R::set_css_property(style, name, &value);
                }
                *prev = value;
                state
            } else {
                (style.clone(), value)
            }
        })
    }

    fn build(self, el: &R::Element) -> Self::State {
        todo!()
    }

    fn rebuild(self, state: &mut Self::State) {}
}

impl<F, C, R> IntoStyle<R> for F
where
    F: Fn() -> C + 'static,
    C: IntoStyle<R> + 'static,
    C::State: 'static,
    R: DomRenderer,
    R::Element: Clone + 'static,
    R::CssStyleDeclaration: Clone + 'static,
{
    type State = Effect<C::State>;

    fn to_html(self, class: &mut String) {
        let value = self();
        value.to_html(class);
    }

    fn hydrate<const FROM_SERVER: bool>(self, el: &R::Element) -> Self::State {
        // TODO FROM_SERVER vs template
        let el = el.clone();
        create_render_effect(move |prev| {
            let value = self();
            if let Some(mut state) = prev {
                value.rebuild(&mut state);
                state
            } else {
                value.hydrate::<FROM_SERVER>(&el)
            }
        })
    }

    fn build(self, el: &R::Element) -> Self::State {
        todo!()
    }

    fn rebuild(self, state: &mut Self::State) {}
}
