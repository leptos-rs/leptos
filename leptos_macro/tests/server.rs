#[cfg(test)]
use cfg_if::cfg_if;

cfg_if! {
    if #[cfg(not(feature = "ssr"))] {
        use leptos::{server, server_fn::{codec, ServerFn}, ServerFnError};
        use std::any::TypeId;

        #[test]
        fn server_default() {
            #[server]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(
                <MyServerAction as ServerFn>::PATH.trim_end_matches(char::is_numeric),
                "/api/my_server_action"
            );
            assert_eq!(TypeId::of::<<MyServerAction as ServerFn>::InputEncoding>(), TypeId::of::<codec::PostUrl>());
        }

        #[test]
        fn server_full_legacy() {
            #[server(FooBar, "/foo/bar", "Cbor", "my_path")]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(<FooBar as ServerFn>::PATH, "/foo/bar/my_path");
            assert_eq!(TypeId::of::<<FooBar as ServerFn>::InputEncoding>(), TypeId::of::<codec::Cbor>());
        }

        #[test]
        fn server_all_keywords() {
            #[server(endpoint = "my_path", encoding = "Cbor", prefix = "/foo/bar", name = FooBar)]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(<FooBar as ServerFn>::PATH, "/foo/bar/my_path");
            assert_eq!(TypeId::of::<<FooBar as ServerFn>::InputEncoding>(), TypeId::of::<codec::Cbor>());
        }

        #[test]
        fn server_mix() {
            #[server(FooBar, endpoint = "my_path")]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(<FooBar as ServerFn>::PATH, "/api/my_path");
            assert_eq!(TypeId::of::<<FooBar as ServerFn>::InputEncoding>(), TypeId::of::<codec::PostUrl>());
        }

        #[test]
        fn server_name() {
            #[server(name = FooBar)]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(
                <FooBar as ServerFn>::PATH.trim_end_matches(char::is_numeric),
                "/api/my_server_action"
            );
            assert_eq!(TypeId::of::<<FooBar as ServerFn>::InputEncoding>(), TypeId::of::<codec::PostUrl>());
        }

        #[test]
        fn server_prefix() {
            #[server(prefix = "/foo/bar")]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(<MyServerAction as ServerFn>::PATH.trim_end_matches(char::is_numeric), "/foo/bar/my_server_action");
            assert_eq!(TypeId::of::<<MyServerAction as ServerFn>::InputEncoding>(), TypeId::of::<codec::PostUrl>());
        }

        #[test]
        fn server_encoding() {
            #[server(encoding = "GetJson")]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(
                <MyServerAction as ServerFn>::PATH.trim_end_matches(char::is_numeric),
                "/api/my_server_action"
            );
            assert_eq!(TypeId::of::<<MyServerAction as ServerFn>::InputEncoding>(), TypeId::of::<codec::GetUrl>());
        }

        #[test]
        fn server_endpoint() {
            #[server(endpoint = "/path/to/my/endpoint")]
            pub async fn my_server_action() -> Result<(), ServerFnError> {
                Ok(())
            }
            assert_eq!(<MyServerAction as ServerFn>::PATH, "/api/path/to/my/endpoint");
            assert_eq!(TypeId::of::<<MyServerAction as ServerFn>::InputEncoding>(), TypeId::of::<codec::PostUrl>());
        }
    }
}
