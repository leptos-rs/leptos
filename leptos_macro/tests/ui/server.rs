use leptos::*;

#[server(endpoint = "my_path", FooBar)]
pub async fn positional_argument_follows_keyword_argument() -> Result<(), ServerFnError> {
    Ok(())
}

#[server(endpoint = "first", endpoint = "second")]
pub async fn keyword_argument_repeated() -> Result<(), ServerFnError> {
    Ok(())
}

#[server(Foo, Bar)]
pub async fn expected_string_literal() -> Result<(), ServerFnError> {
    Ok(())
}
#[server(Foo, Bar, bazz)]
pub async fn expected_string_literal_2() -> Result<(), ServerFnError> {
    Ok(())
}

#[server("Foo")]
pub async fn expected_identifier() -> Result<(), ServerFnError> {
    Ok(())
}

#[server(Foo Bar)]
pub async fn expected_comma() -> Result<(), ServerFnError> {
    Ok(())
}

#[server(FooBar, "/foo/bar", "Cbor", "my_path", "extra")]
pub async fn unexpected_extra_argument() -> Result<(), ServerFnError> {
    Ok(())
}

#[server(encoding = "wrong")]
pub async fn encoding_not_found() -> Result<(), ServerFnError> {
    Ok(())
}

fn main() {}
