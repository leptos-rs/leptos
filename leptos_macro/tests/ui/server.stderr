error: positional argument follows keyword argument
 --> tests/ui/server.rs:3:32
  |
3 | #[server(endpoint = "my_path", FooBar)]
  |                                ^^^^^^

error: keyword argument repeated: `endpoint`
 --> tests/ui/server.rs:8:30
  |
8 | #[server(endpoint = "first", endpoint = "second")]
  |                              ^^^^^^^^

error: expected string literal
  --> tests/ui/server.rs:13:15
   |
13 | #[server(Foo, Bar)]
   |               ^^^

error: expected string literal
  --> tests/ui/server.rs:17:15
   |
17 | #[server(Foo, Bar, bazz)]
   |               ^^^

error: expected identifier
  --> tests/ui/server.rs:22:10
   |
22 | #[server("Foo")]
   |          ^^^^^

error: expected `,`
  --> tests/ui/server.rs:27:14
   |
27 | #[server(Foo Bar)]
   |              ^^^

error: unexpected extra argument
  --> tests/ui/server.rs:32:49
   |
32 | #[server(FooBar, "/foo/bar", "Cbor", "my_path", "extra")]
   |                                                 ^^^^^^^

error: Encoding not found.
  --> tests/ui/server.rs:37:21
   |
37 | #[server(encoding = "wrong")]
   |                     ^^^^^^^
